File: 02/day.kt
fun main() {
	fun getInput(): List<String> {
		return File("input.txt").readText(Charsets.UTF_8).dropLastWhile { it == '\n' }.split('\n')
	}

	fun isValid(nums: List<Int>): Boolean {
		val desc = nums[0] > nums[1]
		for(i in 0 until nums.size - 1) {
			if(((desc && nums[i] > nums[i + 1]) || (!desc && nums[i] < nums[i + 1])) && abs(nums[i] - nums[i + 1]) < 4)
				continue
			return false
		}
		return true
	}

	override fun solve1(): String {
		return getInput().count {
			isValid(it.split(' ').map { n -> n.toInt() })
		}.toString()
	}

	override fun solve2(): String {
		return getInput().count {
			val vals = it.split(' ').map { n -> n.toInt() }
			if(isValid(vals))
				return@count true
			else {
				for(i in vals.indices) {
					val nv = vals.filterIndexed { idx, _ -> idx != i }
					if(isValid(nv))
						return@count true
				}
				return@count false
			}
		}.toString()
	}

	println(solve1())
	println(solve2())
}

File: 02/day.js
import { readFile } from 'node:fs/promises';

const asc = (a, b) => a - b;
const desc = (a, b) => b - a;
const count = (arr, what) => arr.filter(el => el === what).length;
const { abs, sqrt, sin, cos, floor, ceil, round } = Math;

export async function part1() {
	let input = (await readFile('input.txt')).toString().split('\n');
	let ns = [], n = 0;

	for(const line of input) {
		ns = line.split(' ').map(x => +x);
		let dec = ns[0] > ns[1];
		let s = true;
		for(let i = 0; i < ns.length -1; ++i) {
			if(ns[i] > ns[i+1] && dec && abs(ns[i] - ns[i+1]) < 4) {
				continue;
			} else if(ns[i] < ns[i+1] && !dec && abs(ns[i] - ns[i+1]) < 4) {
				continue;
			} else {
				s = false;
				break;
			}
		}
		n += s;
	}
	return n;
}

export async function part2() {
	let input = (await readFile('input.txt')).toString().split('\n');
	let ns = [], n = 0;

	function safe(ns) {
				let dec = ns[0] > ns[1];
		for(let i = 0; i < ns.length -1; ++i) {
			if(ns[i] > ns[i+1] && dec && abs(ns[i] - ns[i+1]) < 4) {
				continue;
			} else if(ns[i] < ns[i+1] && !dec && abs(ns[i] - ns[i+1]) < 4) {
				continue;
			} else {
				return false;
			}
		}
		return true;
	}

	for(const line of input) {
		ns = line.split(' ').map(x => +x);
		console.log(ns);
		if(safe(ns)) {
			++n;
			continue;
		}
		for(let i = 0; i < ns.length; ++i) {
			let a = [...ns];
			a.splice(i, 1);
			if(safe(a)) {
				++n;
				break;
			}
		}
	}
	return n;
}